base_layer/core/src/proof_of_work/difficulty.rs

```diff
@@ -0,0 +1,90 @@
+// Copyright 2019. The Tari Project
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use bitflags::_core::ops::Div;
+use newtype_ops::newtype_ops;
+use serde::{Deserialize, Serialize};
+use std::fmt;
+
+/// Minimum difficulty, enforced in diff retargetting
+/// avoids getting stuck when trying to increase difficulty subject to dampening
+pub const MIN_DIFFICULTY: u64 = 1;
+
+/// The difficulty is defined as the maximum target divided by the block hash.
+#[derive(Debug, Clone, Copy, PartialEq, PartialOrd, Eq, Ord, Deserialize, Serialize)]
+pub struct Difficulty(u64);
+
+impl Difficulty {
+    /// Difficulty of MIN_DIFFICULTY
+    pub fn min() -> Difficulty {
+        Difficulty(MIN_DIFFICULTY)
+    }
+}
+
+impl Default for Difficulty {
+    fn default() -> Self {
+        Difficulty(0)
+    }
+}
+
+// You can only add or subtract Difficulty from Difficulty
+newtype_ops! { [Difficulty] {add sub} {:=} Self Self }
```

Comment:
what sorcery is this!?

Reply:
Magic supplied by the nifty newtype_ops crate :)